### PR TITLE
Fix #3566

### LINKED ISSUE #3566: Missing name when printing type of definition of a record
```agda
postulate
  A : Set

record R : Set where
  field
    a : A
  b : A
  b = a
open R

test = {!b!}
```
Doing `C-c C-d` in the hole above gives the following nonsense
```agda
(: R) → A
```

### PR DESCRIPTION
**Pull request by zgrannan**
This PR attempts to resolve #3566 

I'm not sure if this is the correct approach (as it does not address #434), but it does seem to work for the provided example.

### CODE REVIEW

**Comment by gallais** [comment]
I don't think that would work in general. It does the right thing for guillaume's
original bug report but not for the case with dependencies you use as a test case.
The expression `P (R.a)` you now get in the error message does not make sense:
`R.a` has type `R → A` but `P` expects an `A`, not a function `R → A`! 

What needs to happen is that, because the function type is dependent, we need to
generate a name for `R` and use it later in the telescope. And we should get something
like this: `(r : R) → P (R.a r)`.

Edit: by the way, #434 is a big thing to fix. Your idea of fixing this one without
attempting to fix #434 is a good one.

**Comment by zgrannan** [comment]
@gallais Thanks for the explanation!

I'll try a different approach to address the case of dependent function types defined inside of records.

**Comment by zgrannan** [comment]
@gallais How about this approach? In particular I'm uncertain if the changes to `InternalToAbstract` are correct.

**Comment by jespercockx** [comment]
Looks good to me, thanks for the work!

**Comment by gallais** [comment]
Looked good to me but I built it because I wanted to check that the generated name
is indeed fresh if there is already an `r` in scope. It is the case!

Edit: I should have looked at the test cases, your `Issue3566-3.agda` is already
doing that... :man_facepalming: 

The only thing I am a bit uneasy with is the fact that there are 6 occurrences of
`"r"`. Do they need to all be kept in sync? Should this string be stored directly
in `RecordName` so that it's set in one place & then we simply use that value
everywhere else?

**Comment by zgrannan** [comment]
Ah good point, I changed the name so that it's stored directly in the datatype.

**Review by gallais — APPROVED** [review]

**Comment by gallais** [comment]
Nice. Thanks for the fix!

File: src/full/Agda/TypeChecking/Monad/Base.hs [hunk 1]
```diff
@@ -702,6 +702,11 @@ freshNoName r =
 freshNoName_ :: MonadTCState m => m Name
 freshNoName_ = freshNoName noRange
 
+freshRecordName :: MonadTCState m => m Name
+freshRecordName = do
+  i <- fresh
+  return $ Name i (C.RecordName noRange "r") noRange noFixity'
```
Comment (andreasabel) [inline src/full/Agda/TypeChecking/Monad/Base.hs:708]: I'd prefer "self" as record value name instead of just "r".

**Review by andreasabel — COMMENTED** [review]
Thanks for your work!

This is a clean fix.  However, I am surprised we need to introduce changes in the concrete names.  Concrete names come from parsing, but the record names come from the scope checker or type checker.  I think it would have been enough to change the abstract names to remember this is a record name.

**Pull request merged**